### ISSUE TITLE
parallel errors with different kernels

### ISSUE THREAD

**Issue by rabernat** [ipython/ipython#8169]
I have been using ipython parallel happily for quite a while with python 2.7. I recently upgraded to a setup based on Jupyter Hub / python 3.4 / ipython 3.0.0. The sysadmin for this server set up python 2.7 kernel support in the notebook server, so I can keep running my python-2.7-based notebooks. This setup works great, and I am very impressed with all the new features.

However, I can't get ipython.parallel to work with the 2.7 kernel.

My workflow is as follows:
1. Log in to Jupyter Hub and get my notebook server.
2. Start my cluster through the clusters tab of the notebook server, using the "default" profile.
3. Create a new notebook through the 'new' button, selecting the "Python 2.7" option under "Notebooks". (There is also a "Python 3" option.)
4. Execute some boilerplate parallel code
```python
from IPython.parallel import Client
c = Client()
dview = c.direct_view()
c[:].apply_sync(lambda : "Hello, World")
```

I get the following error from each engine:
```
[0:apply]: 
---------------------------------------------------------------------------TypeError                                 Traceback (most recent call last)/usr/local/anaconda3/lib/python3.4/site-packages/IPython/kernel/zmq/serialize.py in unpack_apply_message(bufs, g, copy)
    160     assert len(bufs) >= 2, "not enough buffers!"
    161     pf = buffer_to_bytes_py2(bufs.pop(0))
--> 162     f = uncan(pickle.loads(pf), g)
    163     pinfo = buffer_to_bytes_py2(bufs.pop(0))
    164     info = pickle.loads(pinfo)
/usr/local/anaconda3/lib/python3.4/site-packages/IPython/utils/codeutil.py in code_ctor(*args)
     22 
     23 def code_ctor(*args):
---> 24     return types.CodeType(*args)
     25 
     26 def reduce_code(co):
TypeError: an integer is required (got type str)
```

This example does work absolutely fine if I select "Python 3" for my notebook in step 3 above.

The obvious hypothesis is that the engines are using the Python 3 kernel, which is incompatible with the Python 2.7 on the client. I assume I have to create a special ipython profile that tells the engines to use a 2.7 kernel. But I can't figure out how to do that.

Here is the information about the default environment (including default Python 3) kernel, from the help --> about menu of my notebook
```
Server Information:

You are using IPython notebook.

The version of the notebook server is 3.0.0-f75fda4 and is running on:
Python 3.4.3 |Anaconda 2.1.0 (64-bit)| (default, Mar  6 2015, 12:03:53) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

Current Kernel Information:

Python 3.4.3 |Anaconda 2.1.0 (64-bit)| (default, Mar  6 2015, 12:03:53) 
Type "copyright", "credits" or "license" for more information.

IPython 3.0.0 -- An enhanced Interactive Python
```

When I am using the 2.7 Kernel, the "Current Kernel Information" is blank. But my versions are:
```python
import IPython
import sys
print( sys.version)
print(IPython.__version__) 
```
```
2.7.9 |Anaconda 2.1.0 (64-bit)| (default, Mar  9 2015, 16:20:48) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
2.2.0
```

**Comment by rabernat** [comment]
Actually, I figured out the solution to this on my own.

created a new profile
```
ipython profile create --parallel --profile=parallel_python2.7
```

Then I edited .ipython/profile_parallel_python2.7/ipcluster_config.py. In this file, there is a configuration option called c.LocalEngineLauncher.engine_cmd. I just had to point this to the correct python executable. In my case:
```
c.LocalEngineLauncher.engine_cmd = ['/usr/local/anaconda/bin/python', '-m', 'IPython.parallel.engine']
```
I did this with all the various engine_cmd and also controller_cmd options. Now it's working great!

Hopefully this issue will be useful documentation if someone else encounters the same problem.
